client/state/data-layer/wpcom/gravatar-upload/index.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * Internal dependencies
+ */
+import {
+	GRAVATAR_UPLOAD_RECEIVE,
+	GRAVATAR_UPLOAD_REQUEST,
+	GRAVATAR_UPLOAD_REQUEST_SUCCESS,
+	GRAVATAR_UPLOAD_REQUEST_FAILURE,
+} from 'state/action-types';
+import { http } from 'state/data-layer/wpcom-http/actions';
+import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';
+import {
+	bumpStat,
+	composeAnalytics,
+	recordTracksEvent,
+	withAnalytics,
+} from 'state/analytics/actions';
+
+export function uploadGravatar( { dispatch }, action ) {
+	const { email, file } = action;
+	dispatch( http( {
+		method: 'POST',
+		path: '/gravatar-upload',
+		body: {},
+		apiNamespace: 'wpcom/v2',
+		formData: [
+			[ 'account', email ],
+			[ 'filedata', file ],
+		],
+	}, action ) );
+}
+
+export function announceSuccess( { dispatch }, { file }, next, data ) {
+	// Return a Promise for testing purposes
```

Comment:
I understand the desire to test everything but this is starting to feel a little much just to be able to test. I'll try and vocalize a few observations:
  - We're already mocking the `FileReader` in the test. We're essentially mocking it here to with the awkward use of the Promise.
  - Handlers are meant to be treated fully asynchronously, so returning a value is definitely going to be adding examples to the code base that will confuse other people reading it.
  - The existing test in the data layer is an integration test, not a unit test - it's testing _both_ the behavior of the `FileReader` _and_ the behavior of the handler.
  - Maybe if we're going to mock it we should full-out mock it in the test which could leave this code in its earlier and simpler form.
  - Could we mock it in such a way that we could manually trigger `load` in the test? `middleware()()(); mockReader.finishLoading(); expect( dispatch ).to…`

Anyway, I'm just wanting to make sure we don't do something we really don't want to do because we think that we have to. In the end, I'd rather have no test than a confusing and fragile one.